parts/tools/gen_diffuser_frame.py

```
@@ -0,0 +1,800 @@
+    adj = _adjacency(segs)
+    caps = _cap_nodes(leds)
+    # a diffuser's own cap welds together every stem that lands on it, so treat
+    # all nodes within the cap footprint of one LED as a single node
+    start = next((k for k in caps if k in adj), None)
+    if start is None:
+        return False
+    seen, stack = {start}, [start]
+    while stack:
+        for nb in adj.get(stack.pop(), ()):
+            if nb not in seen:
+                seen.add(nb); stack.append(nb)
```

Ruff (0.16.0)

[error] Multiple statements on one line (semicolon)

(E702)